src/coreclr/gc/gc.cpp

```diff
@@ -20530,6 +20555,61 @@ void gc_heap::check_and_set_no_gc_oom()
 #endif //MULTIPLE_HEAPS
 }
 
+#ifdef USE_REGIONS
+bool gc_heap::extend_soh_for_no_gc()
+{
+    size_t required = soh_allocation_no_gc;
+    heap_segment* region = generation_allocation_segment (generation_of (0));
```

Comment:
you can start from ephemeral_heap_segment here since it's updated during allocation. generation_allocation_segment is not. there's no need to go back to the beginning of the region list